File: Julia simulations/Main simu.R
# This simulation format should be preferred over the previous one.
rm(list = ls())


# Simulation:
# packages loading ---------------------------------
list.of.packages <- c("tidyverse", "MASS", "stats", "matrixcalc", "stargazer", "foreach", "AER", "doParallel", "ivmodel", "np", "latex2exp", "SteinIV", "magrittr", "weights", "hdm")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
library("tidyverse")
library("MASS")
library("stargazer")
library("foreach")
library("doParallel")
library("AER")
library("matrixcalc")
library("stats")
library("ivmodel")
library("np")
library("latex2exp")
library("SteinIV")
library("magrittr")
library("DEoptim")
library("hdm")     # package for high dimensional models, with the estimator of Belloni, Chen, Chernozukov etc (2012)

# source("Lee_and_Shin.R")   # Donald and Newey (2001) estimator and Lee and Shin (2018) estimator
source("auxiliary functions.R")
source("DGP parameters.R")
# source("P2SLS.R")
# Setup for parallel looping ----------------------------------

parallel:::detectCores()   # detect the number of cores in the CPU of da computa
clust <- makeCluster(parallel:::detectCores()-1)
registerDoParallel(clust )

# Main loop -----------------------------------------
ptm <- proc.time()  
listu <- foreach(s = 1:nsim, .combine = rbind, .packages = c("AER", "MASS", "doParallel", "tidyverse", "np", "hdm"))%:%
  foreach(i = 1:length(VCOV_list) , .combine = rbind)%:%
  # foreach(f = fs_fun, .combine = rbind)%:%
  # foreach( t = type, .combine = rbind)%:%
  foreach( n = N, .combine = rbind) %do%  
  {
    cat(paste("\n","Starting iteration",s,"\n"),
        file="log.txt", append=FALSE)
    
    VCOV <- VCOV_list[[i]]
    # data <- BWdgp( n, beta = param, varz = varz, VCOV, con, polyorder = d )
    data <- dgp(n, parameters = param, varz = 1, VCOV, g = fs_fun, DGP = "normal")
    y <- data$y
    x <- data$x
    z <- data$z1
    
    varu <-  VCOV[1,1]
    varv <-  VCOV[2,2]
    covuv <- VCOV[1,2] 
    varnames <- "x"
    
    
    ols <- lm(y ~ -1 + x )
    betaols <- t(ols$coefficients)
    betaols_var <- rbind(diag(vcov(ols))) 
    tsls <- ivreg(y ~ -1 + x | z )
    beta2sls <- t(tsls$coefficients)
    beta2sls_var <- rbind(diag(vcov(tsls))) 
    
    # map the bandwidth with estimators
    mybeta_ll <- bandrange %>% map_dfr(myest_np, expl = y, endo = x, exo = NULL, instru = z, type = "ll", intercept = FALSE)  # faster than myest
    mybeta_lc <- bandrange %>% map_dfr(myest_np, expl = y, endo = x, exo = NULL, instru = z, type = "lc", intercept = FALSE)
    
    bandwidth <- bandrange
    
    ll <- dplyr::select(mybeta_ll, !ends_with("var"))   # the estimated coefficients
    ll_var <- dplyr::select(mybeta_ll, ends_with("var")) # The estimated variances
    
    # ll_var <- tibble(ll_var = mybeta_ll$varhat) 
    
    lc <- dplyr::select(mybeta_lc, !ends_with("var"))  
    lc_var <- dplyr::select(mybeta_lc, ends_with("var")) # The estimated variances
    
    np <- np_2SLS(Y = y, endo = x, exo = NULL, Z = z, type = "ll", intercept = FALSE)
    np_bhat <- t(np$beta_hat)
    np_varhat <- np$varhat_homoskedasticity#"endo", "endo"]
    opti_h_ll <- np$opti_h
    
    WMD <- wmd( x = x , y = y , z = z, exo = NULL, intercept = FALSE )
    WMDF <- t(WMD$WMDF)
    WMD <- t(WMD$WMD)
    
    iv <- ivmodel::ivmodel(Y = y, D = x , Z = z, beta0 = 1)
    LIML <- as.numeric(ivmodel::LIML(iv, beta0 = 1, alpha = 0.05, heteroSE = FALSE, clusterID = NULL)$point.est)  

    # P2SLS
    # P2SLS <- P2SLS(Y = as.matrix(y), X = dat$dmexfb_alt, Z = dat$dins, K = K, W = as.matrix(W), option = "MSE", weights = dat$popweight, int = TRUE)
  
    # The estimators from Lee and Shin (2019) & Donald and Newey (2001)
    psiz <- poly(data$z1, degree = 10, raw = F, simple = TRUE)
    # # The estimator from Donald and Newey
    DN <- tsls_DN(y = y, x.end = x, z.excl = psiz, method.pre = "one.step", z.pre = psiz)
    DN <- t(DN$bt.hat)
    # # The estimator from Lee and Shin (2018)
    CSA <- tsls_CSA(y = y, x.end = x, z.excl = psiz, method.pre = "one.step", z.pre = psiz)$bt.hat
    CSA <- t(CSA)
    
    # Add rlassoIV by Belloni et al. (2012)
    lassoiv <- rlassoIV(x = psiz, d = x, z = psiz, y = y, select.Z = TRUE)  # Not sure it is correct. Double check
    lasso <- t(lassoiv$coefficients)
    lasso_se <- t(lassoiv$se)
    
    # Change the names of all the estimates
    colnames(betaols) <- paste("ols", varnames, sep = "_")
    colnames(beta2sls) <- paste("tsls", varnames, sep = "_")
    
    colnames(betaols_var) <- paste("ols", varnames, "var", sep = "_")
    colnames(beta2sls_var) <- paste("tsls", varnames, "var", sep = "_")
    
    colnames(ll) <- paste("ll", varnames, sep = "_")
    colnames(lc) <- paste("lc", varnames, sep = "_")
    
    colnames(ll_var) <- paste("ll", varnames, "var", sep = "_")
    colnames(lc_var) <- paste("lc", varnames, "var", sep = "_")    
    
    colnames(np_bhat) <- paste("np", varnames, sep = "_")
    colnames(np_varhat) <- paste("np", varnames, "var", sep = "_")
    
    colnames(WMD) <- paste("WMD", varnames, sep = "_")
    colnames(WMDF) <- paste("WMDF", varnames, sep = "_")    
    
    colnames(DN) <- paste("DN", varnames, sep = "_")
    colnames(CSA) <- paste("CSA", varnames, sep = "_")    
    
    colnames(lasso) <-  paste("Lasso", varnames, sep = "_") 
    colnames(lasso_se) <-  paste("Lasso", varnames, "var", sep = "_")
    
    
    # loo <- loo_choice(x = x, y = y, z = z, h = band, regtype = "ll")
    # bhat_loo <- loo$betahat.betahat
    # bhat_loo_var <- loo$betahat.varhat
    
    # loo <- loo_opti(x, y, z, regtype = "ll")
    # bhat_loo <- loo$betahat.betahat
    # bhat_loo_var <- loo$betahat.varhat
    # loo_h <- loo$opti_h
    # 
    # h_aic <- min_AIC_iv(expl = y, endo = x, instru = z, lower = 0.001, upper = 10, type = "ll", loo = FALSE)
    # aic <-  myest_np(expl = y, endo = x, instru = z, h = h_aic, type = "ll") 
    # bhat_aic <- aic$betahat
    # bhat_aic_var <- aic$varhat
    # psiz <- poly(z, degree = 10, raw = F, simple = TRUE)
    # DN <- tsls_DN(y = y, x.end = x, z.excl = psiz,  method.pre = "one.step", z.pre = psiz)$bt.hat 
    # CSA <- tsls_CSA(y = y, x.end = x, z.excl = psiz, method.pre = "one.step", z.pre = psiz)$bt.hat  
    
    # crit <- bandrange %>% map_dbl(iv_AIC, expl = y, endo = x, instru = z, type = t, loo = FALSE)
    # h_opti <- bandrange[which.min(crit)]
    # AIC_est <- myest(expl = y, endo = x, instru = z, h = h_opti, type = t, cv = FALSE)
    # 
    # # crit_loo <- bandrange %>% map_dbl(iv_AIC, expl = y, endo = x, instru = z, type = t, loo = TRUE)
    # # h_opti_loo <- bandrange[which.min(crit_loo)]
    # # AIC_est_loo <- myest(expl = y, endo = x, instru = z, h = h_opti_loo, type = t, cv = TRUE)
    # 
    # iv_AIC(expl = y, endo = x, instru = z, bandwidth = 0.01, type = "ll", loo = FALSE)
    
    # sol_ll <- min_AIC_iv(expl = y, endo = x, instru = z, lower = n^(-1), upper = 100, type = "ll", loo = FALSE)
    # hstar_ll <- sol_ll$minimizer
    # minimum_ll <- sol_ll$minimum    
    # AIC_opti_est_ll <- myest_np(expl = y, endo = x, instru = z, h = hstar_ll, type = "ll")$betahat
    # 
    # sol_lc <- min_AIC_iv(expl = y, endo = x, instru = z, lower = n^(-1), upper = 100, type = "lc", loo = FALSE)
    # hstar_lc <- sol_lc$minimizer
    # minimum_lc <- sol_lc$minimum    
    # AIC_opti_est_lc <- myest_np(expl = y, endo = x, instru = z, h = hstar_lc, type = "lc")$betahat
    
    print(s)
    return(data.frame(simu = s, sample = n, varu = varu, varv = varv, covuv = covuv, 
                      fs_fun = fs_fun,
                      betaols,
                      betaols_var,
                      beta2sls,
                      beta2sls_var,
                      bandwidth = bandrange,
                      ll,
                      ll_var,
                      lc,
                      lc_var,
                      np_bhat,
                      np_varhat,
                      WMD,
                      WMDF,
                      LIML,   #,
                      CSA,
                      DN,
                      lasso,
                      lasso_se
                      
    ))
    print(s)
  }
proc.time() - ptm 
save.image() 
message <- geterrmessage()
cat(paste("\n","Last error", message,"\n"),
    file="errors.txt", append=FALSE)
name_file <- paste("kernel 2SLS, ", fs_fun, " design", ".RData", sep = "")
# name_file <- paste("kernel 2SLS simulation, ", "all fs_fun", ".RData", sep = "")
save(listu , file = name_file)

# Results ----------------------------

# The most recent results (nsim = 10,000 and all designs)
# load("kernel 2SLS simulation, all fs_fun.RData")
# 
# # The most recent results with separate designs
# load("kernel 2SLS, A&L DGP.RData")
# load("kernel 2SLS, linear DGP.RData")
# load("kernel 2SLS, log DGP.RData")
# 
# # The results in the respective folders
# load("Simulation results/linear design/Alternative kernel 2SLS consistency, linear DGP.RData")
# load("Simulation results/A&L design/Alternative kernel 2SLS consistency, A&L DGP.RData")
# load("Simulation results/doppler + dnorm design/Alternative kernel 2SLS consistency, dnorm, doppler DGP.RData")
# load("Simulation results/binary x design/Alternative kernel 2SLS consistency, binary_x DGP.RData")
# load("Simulation results/binary x design/Alternative kernel 2SLS consistency, binary_x DGP (newer).RData")   # for power graphs
# load("Simulation results/log design/Alternative kernel 2SLS consistency, log DGP.RData")
# 
# # The results from the Julia simulations
# listu <- read_csv("first results (most recent).csv")
# listu <- read_csv("first results (most recent, Dec 6).csv")
listu <- read_csv("first results (Dec 13).csv")



MC_b <- listu[, -1] %>% dplyr::select(!ends_with("var")) %>%
  group_by( sample, covuv, fs_fun, bandwidth ) %>% 
  summarise_all (MSE, parameter = param, type = "bias" ) 

MC_v <- listu[, -1] %>% dplyr::select(!ends_with("var")) %>%
  group_by( sample, covuv, fs_fun, bandwidth ) %>% 
  summarise_all (MSE, parameter = param, type = "variance" )

MC_MSE <- listu[, -1] %>% dplyr::select(!ends_with("var")) %>%
  group_by( sample, covuv, fs_fun, bandwidth ) %>% 
  summarise_all (MSE, parameter = param, type = "MSE" )

MC_medAD <- listu[, -1] %>% dplyr::select(!ends_with("var")) %>%
  group_by( sample, covuv, fs_fun, bandwidth ) %>% 
  summarise_all (MSE, parameter = param, type = "medAD" )

MC_medbias <- listu[, -1] %>% dplyr::select(!ends_with("var")) %>%
  group_by( sample, covuv, fs_fun, bandwidth ) %>% 
  summarise_all (MSE, parameter = param, type = "medbias" )

# MC_b <- listu[, -1] %>% dplyr::select(!ends_with("var")) %>%
#   group_by( sample, varu, varv, covuv, fs_fun, bandwidth ) %>% 
#   summarise_all (MSE, parameter = param, type = "bias" ) 
# 
# MC_v <- listu[, -1] %>% dplyr::select(!ends_with("var")) %>%
#   group_by( sample, varu, varv, covuv, fs_fun, bandwidth ) %>% 
#   summarise_all (MSE, parameter = param, type = "variance" )
# 
# MC_MSE <- listu[, -1] %>% dplyr::select(!ends_with("var")) %>%
#   group_by( sample, varu, varv, covuv, fs_fun, bandwidth ) %>% 
#   summarise_all (MSE, parameter = param, type = "MSE" )
# 
# MC_medAD <- listu[, -1] %>% dplyr::select(!ends_with("var")) %>%
#   group_by( sample, varu, varv, covuv, fs_fun, bandwidth ) %>% 
#   summarise_all (MSE, parameter = param, type = "medAD" )
# 
# MC_medbias <- listu[, -1] %>% dplyr::select(!ends_with("var")) %>%
#   group_by( sample, varu, varv, covuv, fs_fun, bandwidth ) %>% 
#   summarise_all (MSE, parameter = param, type = "medbias" )

# Graphs ---------------------
# MC_b <- MC_b %>% dplyr::select(!contains("W1") & !contains("W2"))
# MC_v <- MC_v %>% dplyr::select(!contains("W1") & !contains("W2"))
# MC_MSE <- MC_MSE %>% dplyr::select(!contains("W1") & !contains("W2"))


# Not needed if simulations come from Julia
MC_b <- MC_b %>% rename(ols = ols_x,
                        tsls = tsls_x,
                        ll = ll_x,
                        lc = lc_x,
                        np_tsls = np_x,
                        WMD = WMD_x,
                        WMDF = WMDF_x,
                        DN = DN_x,
                        CSA = CSA_x
                        #,
                        # ptsls = ptsls     
)
MC_v <- MC_v %>% rename(ols = ols_x,
                        tsls = tsls_x,
                        ll = ll_x,
                        lc = lc_x,
                        np_tsls = np_x,
                        WMD = WMD_x,
                        WMDF = WMDF_x,
                        DN = DN_x,
                        CSA = CSA_x#,
                        # ptsls = ptsls     
)
MC_MSE <- MC_MSE %>% rename(ols = ols_x,
                            tsls = tsls_x,
                            ll = ll_x,
                            lc = lc_x,
                            np_tsls = np_x,
                            WMD = WMD_x,
                            WMDF = WMDF_x,
                            DN = DN_x,
                            CSA = CSA_x#,
                            # ptsls = ptsls     
)

MC <- list(MC_b, MC_v, MC_MSE)
# If looking at different bias measures
# MC <- list(MC_b, MC_medbias, MC_medAD)
# unique_designs <- unique(MC_MSE[, 1:5])    # As the same designs show up multiple times due to the different values of the bandwidth, I report the unique designs
unique_designs <- unique(MC_MSE[, 1:3])    # The Julia simulation does not have varu and varv

1:nrow(unique_designs) %>% map(MC_graph, MC = MC, exo = FALSE)

# Relative bias, variance etc... -------------------

relative_b <- MC_b %>% dplyr::select(!ends_with("var") & !bandwidth & !contains("_W") & !ll_x & !lc_x) %>%
  group_by( sample, varu, varv, covuv, fs_fun) %>%
  summarise_all(mean) %>%   # all the numbers are constant for a same design, so the mean does not change anything. It allows to regroup everything to avoid repeating design because of previous bandwidths values
  mutate(ols = ols_x/tsls_x,
         liml = LIML/tsls_x,
         WMDF = WMDF/tsls_x,
         np_x = np_x/tsls_x,
         # ll_x = ll_x/tsls_x,
         # lc_x = lc_x/tsls_x,
         tsls_x = tsls_x/tsls_x  ) 


relative_v <- MC_v %>% dplyr::select(!ends_with("var") & !bandwidth & !contains("_W")) %>%
  group_by( sample, varu, varv, covuv, fs_fun) %>%
  summarise_all(mean) %>%   # all the numbers are constant for a same design, so the mean does not change anything. It allows to regroup everything to avoid repeating design because of previous bandwidths values
  mutate(ols = ols_x/tsls_x,
         liml = LIML/tsls_x,
         WMDF = WMDF/tsls_x,
         np_x = np_x/tsls_x,
         ll_x = ll_x/tsls_x,
         lc_x = lc_x/tsls_x,
         tsls_x = tsls_x/tsls_x  ) 


relative_MSE <- MC_MSE %>% dplyr::select(!ends_with("var") & !bandwidth & !contains("_W")) %>%
  group_by( sample, varu, varv, covuv, fs_fun) %>%
  summarise_all(mean) %>%   # all the numbers are constant for a same design, so the mean does not change anything. It allows to regroup everything to avoid repeating design because of previous bandwidths values
  mutate(ols = ols_x/tsls_x,
         liml = LIML/tsls_x,
         WMDF = WMDF/tsls_x,
         np_x = np_x/tsls_x,
         ll_x = ll_x/tsls_x,
         lc_x = lc_x/tsls_x,
         tsls_x = tsls_x/tsls_x  ) 

relative_medbias <- MC_medbias %>% dplyr::select(!ends_with("var") & !bandwidth & !contains("_W")) %>%
  group_by( sample, varu, varv, covuv, fs_fun) %>%
  summarise_all(mean) %>%   # all the numbers are constant for a same design, so the mean does not change anything. It allows to regroup everything to avoid repeating design because of previous bandwidths values
  mutate(ols = ols_x/tsls_x,
         liml = LIML/tsls_x,
         WMDF = WMDF/tsls_x,
         np_x = np_x/tsls_x,
         ll_x = ll_x/tsls_x,
         lc_x = lc_x/tsls_x,
         tsls_x = tsls_x/tsls_x  ) 

relative_medAD <- MC_medAD %>% dplyr::select(!ends_with("var") & !bandwidth & !contains("_W")) %>%
  group_by( sample, varu, varv, covuv, fs_fun) %>%
  summarise_all(mean) %>%   # all the numbers are constant for a same design, so the mean does not change anything. It allows to regroup everything to avoid repeating design because of previous bandwidths values
  mutate(ols = ols_x/tsls_x,
         liml = LIML/tsls_x,
         WMDF = WMDF/tsls_x,
         np_x = np_x/tsls_x,
         ll_x = ll_x/tsls_x,
         lc_x = lc_x/tsls_x,
         tsls_x = tsls_x/tsls_x  ) 

# Save the relative stats tables in .csv format 
design <- "log"
stats <- c("relative_b", "relative_v", "relative_MSE", "relative_medbias", "relative_medAD")
set <- list(relative_b, relative_v, relative_MSE, relative_medbias, relative_medAD)  
for (i in 1:length(set))
{
  name <- paste(design, "_", stats[i], ".csv", sep = "")
  write_csv(set[[i]], name)
}

# Power graphs-------------------------------------

# unique_designs <- unique(listu[, c(2:6)])    # As the same designs show up multiple times due to the different values of the bandwidth, I report the unique designs
unique_designs <- unique(listu[, c(2:4)])    # In Julia

grid <- seq(-1, 3, 0.1) # grid centered around the true parameter beta = 1
# Saves a bunch of power graphs according to each design
1:nrow(unique_designs) %>% map( power_graph, data = listu, H0_range = grid, level = 0.05, by = "sample")
1:nrow(unique_designs) %>% map( power_graph, data = listu, H0_range = grid, level = 0.05, by = "bandwidth")

File: Julia simulations/Main simu.jl
# This script is a test on some data, to compare if my functions in Julia and R do the same thing
# Pkg.add("Nemo")
# Pkg.add("Polynomials")
# Pkg.add("SharedArrays")
# Pkg.add("Distributed")
# Pkg.add("StatsModels")

using Polynomials
#using StatsModels
using DataFrames
using Distributions
#using KernelEstimator
#using BlackBoxOptim
using LinearAlgebra
using Optim
using CSV
#using NLsolve
using SharedArrays      # Used to create the matrix the simulation will fill
#using Distributed

include("auxiliary functions.jl")
include("parameters.jl")
include("Estimators.jl")

# Different covariances for the
covs = [0.1, 0.5, 0.9]
# Number of estimators used in the simulation (x2 since we include the standard errors. WMDF doesn't include the standard error... yet)
n_estimators = 14 + 1
# how many bandwidth columns we report
band = 5  # I want to report all the bandwidths: the one we loopover, 1 from using np, 1 from my loo criterion, and 2 coming from the AIC criteria
mat = SharedArray(zeros(nsim*length(bandrange), (4 + n_estimators + band), length(N), length(SIGMA), length(fs_fun)) )
start = time()

# Threads.@threads
#@time begin
# @distributed
 for i = 1:nsim
        for j = 1:length(SIGMA)
        for n in 1:length(N)
		for f in 1:length(fs_fun)
    data = dgp(N[n], beta_0, mu, SIGMA[j], design = fs_fun[f])
    # Got loop over the bandwidth here to keep the same data set throughout

    OLS = ols(data.x, data.y, intercept = "false")
    TSLS = tsls(data.x, data.y, data.z, intercept = "false")
#    bw = myest_bw(data.x, data.y, data.z; regtype = "lc")

    bw_np = myest_bw(data.x, data.y, data.z; regtype = "ll").bandwidth[1]
    np_2SLS = myest(data.x, data.y, data.z, bw_np; regtype = "ll")

# AIC version for the IV case
aic_bw = min_AIC_iv( data.y, data.x, data.z; lower = 0.01, upper = 1000, type = "ll").opt_h
aic =  myest(data.x, data.y, data.z, aic_bw; regtype = "ll")

# AIC version on the first stage
aic_bw_stage1 = min_AIC( data.x, data.z; lower = 0.01, upper = 1000, type = "ll").opt_h
aic_stage1 =  myest(data.x, data.y, data.z, aic_bw_stage1; regtype = "ll")

# WMDF
wmd = wmdf(data.x, data.y, data.z; intercept = "false").WMDF


# Add that to the simulation
kernel_TV = loo_opti(data.x, data.y, data.z, regtype = "ll")
bw_TV = kernel_TV.Bandwidth[1]
# Maybe for later, the whole formatting thing is still tricky
#    kernel_TV_h = map(h -> myest( data.x, data.y, data.z, h, regtype = "lc"), bandrange )
#    gna = groupby(kernel_TV_h, :bandwidth)
#    combine(gna), :estimate => mean)

     # NP = np_tsls(data.x, data.y, data.z, option = locallinear)
     # LIML = liml(data.x, data.y, data.z)
     # WMDF = wmdf(data.x, data.y, data.z; intercept = "false").WMDF
     # res_lc = TV_kernel_iv(data.y, data.x, data.z, option = "lc")
     # res_ll = TV_kernel_iv(data.y, data.x, data.z, option = "ll")
     # TV_lc = res_lc.bt_hat
     # TV_ll = res_ll.bt_hat
     # opt_h_lc = res_lc.opt_h
     # opt_h_ll = res_ll.opt_h
     # TV_lc_h = map(bw -> kernel_tv( data.x, data.y, data.z, "lc", bw), bandrange )
     # TV_ll_h = map(bw -> kernel_tv( data.x, data.y, data.z, "ll", bw), bandrange )
     bhat_ols = OLS.Estimates[1]
     bhat_tsls = TSLS.Estimates[1]
     bhat_np = np_2SLS.Estimates[1]
	 bhat_TV = kernel_TV.Estimates[1]
     bhat_aic = aic.Estimates[1]
     bhat_aic_stage1 = aic_stage1.Estimates[1]
     bhat_wmdf = wmd

     bhat_ols_se = OLS."Standard errors"[1]
     bhat_tsls_se = TSLS."Standard errors"[1]
     bhat_np_se = np_2SLS."Standard errors"[1]
     bhat_TV_se = kernel_TV."Standard errors"[1]
     bhat_aic_se = aic."Standard errors"[1]
     bhat_aic_stage1_se = aic_stage1."Standard errors"[1]



# Loop over the bandwidths
     for h in 1:length(bandrange)
         np_2SLS_h = myest(data.x, data.y, data.z, bandrange[h]; regtype = "ll")
         bhat_np_h = np_2SLS_h.Estimates[1]
         bhat_np_h_se = np_2SLS_h."Standard errors"[1]

    mat[(i-1)*length(bandrange) + h, 1, n, j, f] = i
	mat[(i-1)*length(bandrange) + h, 2, n, j, f] = N[n]
    mat[(i-1)*length(bandrange) + h, 3, n, j, f] = covs[j]
	mat[(i-1)*length(bandrange) + h, 4, n, j, f] = f
	mat[(i-1)*length(bandrange) + h, 5, n, j, f] = bhat_ols
    mat[(i-1)*length(bandrange) + h, 6, n, j, f] = bhat_ols_se

    mat[(i-1)*length(bandrange) + h, 7, n, j, f] = bhat_tsls
    mat[(i-1)*length(bandrange) + h, 8, n, j, f] = bhat_tsls_se

    mat[(i-1)*length(bandrange) + h, 9, n, j, f] = bandrange[h]

    mat[(i-1)*length(bandrange) + h, 10, n, j, f] = bhat_np_h
    mat[(i-1)*length(bandrange) + h, 11, n, j, f] = bhat_np_h_se


    mat[(i-1)*length(bandrange) + h, 12, n, j, f] = bhat_np
    mat[(i-1)*length(bandrange) + h, 13, n, j, f] = bhat_np_se

	mat[(i-1)*length(bandrange) + h, 14, n, j, f]= bhat_TV
    mat[(i-1)*length(bandrange) + h, 15, n, j, f]= bhat_TV_se

    mat[(i-1)*length(bandrange) + h, 16, n, j, f]= bhat_aic
    mat[(i-1)*length(bandrange) + h, 17, n, j, f]= bhat_aic_se

    mat[(i-1)*length(bandrange) + h, 18, n, j, f]= bhat_wmdf

    mat[(i-1)*length(bandrange) + h, 19, n, j, f]= bhat_aic_stage1
    mat[(i-1)*length(bandrange) + h, 20, n, j, f]= bhat_aic_stage1_se

    mat[(i-1)*length(bandrange) + h, 21, n, j, f]= aic_bw
    mat[(i-1)*length(bandrange) + h, 22, n, j, f]= aic_bw_stage1
    mat[(i-1)*length(bandrange) + h, 23, n, j, f]= bw_np
    mat[(i-1)*length(bandrange) + h, 24, n, j, f]= bw_TV


end
end
end
end
end

#end

elapsed = time() - start
# On nsim = 2
# no parallel: 3.4 sec
# using @distributed:

# Results formatting time
ncol = size(mat, 2)

mata = Array{Float64}(undef, 1, ncol)   # Make a row to concatenate the results vertically
for i in 1:length(fs_fun), j in 1:length(SIGMA), k in 1:length(N)
global mata = vcat(mata, mat[:,:, k, j, i])   # The "global" is needed on recent versions of Julia since otherwise, mata is reassigned every iteration and Julia does not recognize it
end
mat = mata[2:end,:]
#mat = vcat(mat[:,:,:, 1], mat[:,:,:, 2], mat[:,:,:, 3])
#mat = vcat(mat[:, :, 1], mat[:, :, 2])

results = DataFrame( mat , :auto)
results[!, :x4] = Int64.(results[:, :x4])
results[!, :x4] = fs_fun[results[:, 4]]
#ll_h_names = map(k -> name_k(k, "ll_h"), bandrange)
#lc_h_names = map(k -> name_k(k, "lc_h"), bandrange)

# To do:
# - Change names to match the R simulations
# - Add WMDF
# - Add the variances
# - Double check the cross validation for betahat formula
# - Rename current results
col_names = ["s", "sample", "covuv", "fs_fun",
             "ols",
             "ols_se",
             "tsls",
             "tsls_se",
             "bandwidth",
			 "ll",
             "ll_se",
            # "lc",
            # "lc_se"
             "np_tsls",
             "np_tsls_se",
             "kernel_TV",
             "kernel_TV_se",
             "bhat_aic",
             "bhat_aic_se",
             "WMDF",
             "bhat_aic_stage1",
             "bhat_aic_stage1_se",
             "bw_aic",
             "bw_aic_stage1",
             "bw_np",
             "bw_TV"
              ]
#col_names = vcat(col_names, ll_h_names, lc_h_names)
#names!(results, [Symbol("$i") for i in col_names ])
results = rename!(results, col_names)
results

#writedlm("caca.csv", bet, delim = ',')
CSV.write("F:/Kernel 2SLS/Julia simulations/first results (unfinished).csv", results)
